core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -322,6 +318,24 @@ private boolean pollForCompletion(BlockingQueue<CompletableFuture<?>> queue, lon
       return previousFuture;
    }
 
+   // Unfortunately a tx exception is part of a caused and suppressed - we need to make sure the real cause is available
```

Comment:
Please replace the comment with one or two actual examples so we'll know what to expect when debugging.
And are you sure the `TimeoutException` is always the most nested suppressed throwable, or should we check if any of the suppressed exceptions is a `TimeoutException`?

Reply:
In my testing it was always the most nested. I can try to debug it again to get the layout, it was something like CompletionException caused by XAException supressing RemoteException caused by TimeoutException or something like that :P

Reply:
I'm quite curious now why the `RemoteException` is suppressed... looking at `TransactionCoordinator.prepare()`, it calls `initCause()`, not `addSuppressed`.

Reply:
I reproduced again; I missed the ordering and the RollbackException :)

```
CompletionException
 -> caused by RollbackException
    -> suppressing XAException
       -> caused by RemoteException
         -> caused by TimeoutException
```

Reply:
adding to comments

Reply:
Do we still need the comment here, now that it's explained at the call site?

And I really hope lock timeouts aren't going to be very common, because these exceptions aren't cheap...

Off-topic, but definitely related, I don't see any special handling for `WriteSkewException`, so I wonder if the user is supposed to go through the same hoops in order to detect a WSE and retry the tx. I'm assuming we can't avoid the `RollbackException`, at least for explicit transactions, but it would be nice if the user could check for WSE with

```
catch (RollbackException e) {
   if (e.getCause() instanceof WriteSkewException) {
       ... retry ... 
   }
}
```

Reply:
> Do we still need the comment here, now that it's explained at the call site?

I guess I can remove it, it seemed pretty mundane to me.

>And I really hope lock timeouts aren't going to be very common, because these exceptions aren't cheap...

The LockTimeout should be very rare. The tests we have just pound expiration really hard, as it keeps expiring and calling get/replace within milliseconds 100 times.

> Off-topic, but definitely related, I don't see any special handling for `WriteSkewException`, so I wonder if the user is supposed to go through the same hoops in order to detect a WSE and retry the tx. I'm assuming we can't avoid the `RollbackException`, at least for explicit transactions, but it would be nice if the user could check for WSE with
> 
> ```
> catch (RollbackException e) {
>    if (e.getCause() instanceof WriteSkewException) {
>        ... retry ... 
>    }
> }
> ```

Are you saying in general or in this case? Cause I wouldn't expect to handle `WriteSkewException` here and just propagate it. In the general, case I agree it would be quite annoying for a user to see the `WriteSkewException` as it is buried a few layers deep, including a suppressed (which is a royal PITA).

Reply:
I meant in general, I don't expect users to call `removeXExpired()`, but I debugged one of our tests and the hierarchy looks exactly the same for `WriteSkewException`. I created https://issues.redhat.com/browse/ISPN-11167 for the `WriteSkewException`.